File: src/sys/pkg/lib/omaha-client/src/metrics/mock.rs
// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use crate::metrics::{Metrics, MetricsReporter};
use anyhow::{format_err, Error};

/// A mock implementation of MetricsReporter save the metrics into a vector.
#[derive(Debug)]
pub struct MockMetricsReporter {
    pub should_fail: bool,
    pub metrics: Vec<Metrics>,
}

impl MockMetricsReporter {
    pub fn new_failing() -> Self {
        MockMetricsReporter { should_fail: true, metrics: vec![] }
    }

    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        MockMetricsReporter { should_fail: false, metrics: vec![] }
    }
}

impl MetricsReporter for MockMetricsReporter {
    fn report_metrics(&mut self, metrics: Metrics) -> Result<(), Error> {
        self.metrics.push(metrics);
        if self.should_fail {
            Err(format_err!("should_fail is true"))
        } else {
            Ok(())
        }
    }
}

mod tests {
    use super::*;
    use std::time::Duration;

    #[test]
    fn test_mock_metrics_reporter() {
        let mut mock = MockMetricsReporter::new();
        let result = mock.report_metrics(Metrics::UpdateCheckResponseTime {
            response_time: Duration::from_secs(2),
            successful: true,
        });
        assert!(result.is_ok(), "{:?}", result);
    }

    #[test]
    fn test_mock_metrics_reporter_error() {
        let mut mock = MockMetricsReporter::new_failing();
        let result = mock.report_metrics(Metrics::UpdateCheckResponseTime {
            response_time: Duration::from_secs(5),
            successful: true,
        });
        assert!(result.is_err());
    }
}

File: tools/testing/testrunner/mock.go
// Copyright 2021 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package testrunner

import (
	"context"
	"io"

	"go.fuchsia.dev/fuchsia/tools/lib/ffxutil"
)

type MockFFXTester struct {
	CmdsCalled []string
}

func (f *MockFFXTester) SetStdoutStderr(_, _ io.Writer) {
}

func (f *MockFFXTester) run(cmd string) error {
	f.CmdsCalled = append(f.CmdsCalled, cmd)
	return nil
}

func (f *MockFFXTester) Test(_ context.Context, _ []ffxutil.TestDef, _ string, _ ...string) (*ffxutil.TestRunResult, error) {
	f.run("test")
	return &ffxutil.TestRunResult{Outcome: ffxutil.TestPassed}, nil
}

func (f *MockFFXTester) Snapshot(_ context.Context, _, _ string) error {
	return f.run("snapshot")
}

func (f *MockFFXTester) Stop() error {
	return f.run("stop")
}

func (f *MockFFXTester) ContainsCmd(cmd string) bool {
	for _, c := range f.CmdsCalled {
		if c == cmd {
			return true
		}
	}
	return false
}
